### PR TITLE
[mlir-hlo] Added BufferReuse Optimization. 

### PR DESCRIPTION
**Pull request by ghost**
In this PR, we want to introduce a new optimization on reusing already allocated buffers to save memory consumption. The optimization consists of two steps. First, for each buffer, we find a list of buffers that are potential reuses. A possible reuse has the following properties: 
- the types are compatible
- no interference in the `UserangeAnalysis` (#48847)
- the dominance is still given

The second step is a fixpoint iteration over the potential reuses. This step is divided into two substeps: 
- try to assign possible reuses for each buffer
- update the potential reuses based on the assignments from step 1.
After the distribution of all possible reusable buffers is done, they are actually replaced.
```
For example:                        Result:
func @simpleReuse(%arg0: i1) {      func @simpleReuse(%arg0: i1) {
  %0 = alloc()                        %0 = alloc()
  %1 = alloc()
  cond_br %arg0, ^bb1, ^bb2           cond_br %arg0, ^bb1, ^bb2
 ^bb1:                               ^bb1:
  use(%0)                             use(%0)
  br ^bb3                             br ^bb3
 ^bb2:                               ^bb2:
  use(%1)                             use(%0)
  br ^bb3                             br ^bb3
 ^bb3:                               ^bb3:
  return                              return
}                                   }
```
In this simple example `%1` can be replaced with `%0`, because all requirements mentioned above are fulfilled.

This PR is a follow up to #48847, in which we introduced the `UserangeAnalysis`.

### CODE REVIEW

**Review by sanjoy — CHANGES_REQUESTED** [review]
Should this be split into multiple PRs, one with just the Userange analysis and one with the buffer_reuse optimization?

**Comment by gbaned** [comment]
@dfki-albo  Can you please check @sanjoy's comments and keep us posted ? Thanks!

File: tensorflow/compiler/mlir/hlo/include/mlir-hlo/Analysis/userange_analysis.h [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+  /// Checks each operand inside the operation for its memory effects and
+  /// separates them into read and write. Operands with read effects are added
+  /// to the opToReadMap
+  void
+  getMemoryEffects(Operation *op,
+                   DenseMap<Operation *, SmallPtrSet<Value, 2>> &opToReadMap);
+
+  /// Computes the ID for the operation. If the operation contains operands
+  /// which have read effects, the returning ID will be odd.
```
[the remark this answers is not in the document]
Comment (dfki-albo) [reply tensorflow/compiler/mlir/hlo/include/mlir-hlo/Analysis/userange_analysis.h:76]: Yes, you are right. This part is still work in progress.  I will change the PR of the UserangeAnalysis to a draft and work on this issue. Also I tried to split this PR into 2 parts. Part 1 can be found here: [#48847](https://github.com/tensorflow/tensorflow/pull/48847).

File: tensorflow/compiler/mlir/hlo/lib/Analysis/userange_analysis.cc [hunk 2]
```diff
@@ -0,0 +1,540 @@
...
+/// Note: This assumes that all intervals of b are included in some interval
+///       of a.
+void UserangeAnalysis::intervalSubtract(IntervalVector& a,
+                                        const IntervalVector& b) const {
+  auto iterB = b.begin();
+  auto endB = b.end();
+  for (auto iterA = a.begin(), endA = a.end();
+       iterA != endA && iterB != endB;) {
+    // iterA is strictly before iterB => increment iterA.
+    if (iterA->second < iterB->first) ++iterA;
```
[the remark this answers is not in the document]
Comment (dfki-albo) [reply tensorflow/compiler/mlir/hlo/lib/Analysis/userange_analysis.cc:469]: This is not possible. The prerequisite is that B is a "proper subset" of A. If `iterA` is on the last element and `IntervalVector b` is not subtracted yet, then `iterA` cannot be before `iterB`.

File: tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+      if (currentReuserSet.empty()) break;
+
+      // Step 2 of the fixpoint iteration: Update the potentialReuseVectors for
+      // each value in the potentialReuseMap. Due to the chosen replacements in
+      // step 1 some values might not be replaceable anymore. Also remove all
+      // replaced values from the potentialReuseMap.
+      for (auto itReuseMap = potentialReuseMap.begin();
+           itReuseMap != potentialReuseMap.end();) {
+        Value item = itReuseMap->first;
+        SmallVector<Value, 4>* potReuses = &itReuseMap->second;
```
Comment (sherhut) [inline tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc:148]: How about using a reference?

File: tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+        // Get the defining block of itemA.
+        Block* defOpBlock = itemA.isa<BlockArgument>()
+                                ? itemA.getParentBlock()
+                                : itemA.getDefiningOp()->getBlock();
+
+        // The defining block of itemA has to dominate all uses of itemB.
+        if (!dominatesAllUses(defOpBlock, itemB)) continue;
+
+        // Insert itemB into the right place of the potReuseVector. The order of
```
Comment (sherhut) [inline tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc:65]: It would be interesting to explore other orders here. This is essentially greedy but one could try prioritize reuses for certain cases. Like if one has a copy, then source and destination should preferably be reused.

File: tensorflow/compiler/mlir/hlo/tools/mlir-hlo-opt/mlir-hlo-opt.cpp [hunk 5]
```diff
@@ -26,6 +27,7 @@ int main(int argc, char **argv) {
   mlir::registerAllPasses();
   mlir::mhlo::registerAllMhloPasses();
   mlir::lmhlo::registerAllLmhloPasses();
+  mlir::registerAllTransformPasses();
```
Comment (sherhut) [inline tensorflow/compiler/mlir/hlo/tools/mlir-hlo-opt/mlir-hlo-opt.cpp:31]: Why are these needed here and everywhere? Was that for debugging?

File: tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+      SmallVector<Value, 4> potReuseVector;
+      for (BufferPlacementAllocs::AllocEntry entry : allocs) {
+        Value itemB = std::get<0>(entry);
+        // Do not compare an item to itself and make sure that the value of item
+        // B is not a BlockArgument. BlockArguments cannot be reused. Also
+        // perform a type check.
+        if (itemA == itemB || !checkTypeCompatibility(itemA, itemB)) continue;
+
+        // Check if itemA can replace itemB.
+        if (!userange.rangesInterfere(itemA, itemB)) continue;
```
[the remark this answers is not in the document]
Comment (dfki-albo) [reply tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc]: Yes, you are right. I'll add this to the comment and also move the negation into the method, because otherwise the name is misleading.

**Comment by gbaned** [comment]
@dfki-albo Can you please resolve conflicts? Thanks!

File: tensorflow/compiler/mlir/hlo/include/mlir-hlo/Transforms/passes.td [hunk 7]
```diff
@@ -18,6 +18,17 @@ limitations under the License.
 
 include "mlir/Pass/PassBase.td"
 
+def BufferReuse : FunctionPass<"buffer-reuse"> {
+  let summary = "Reuses already allocated buffers to save allocation "
+                "operations if is provably safe.";
```
Comment (sherhut) [inline tensorflow/compiler/mlir/hlo/include/mlir-hlo/Transforms/passes.td:23]: nit: if it is

File: tensorflow/compiler/mlir/hlo/lib/Analysis/test_userange_analysis.cc [hunk 8]
```diff
@@ -30,6 +30,11 @@ struct TestUserangePass : public TestUserangeBase<TestUserangePass> {
     registry.insert<mlir::lmhlo::LmhloDialect>();
   }
 
+  StringRef getArgument() const final { return "test-print-userange"; }
```
Comment (sherhut) [inline tensorflow/compiler/mlir/hlo/lib/Analysis/test_userange_analysis.cc:33]: Why is this needed?

File: tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+        if (itemA == itemB || !checkReuseCompatibility(itemA, itemB))
+          continue;
+
+        // Check if itemA interferes with itemB. If this is the case no reuse is
+        // possible.
+        if (userange.rangesInterfere(itemA, itemB))
+          continue;
+
+        // Get the defining block of itemA.
+        Block *defOpBlock = itemA.isa<BlockArgument>()
```
Comment (sherhut) [inline tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc:59]: You can always use `getParentBlock()` here. It returns the block of the defining op, as well.

Comment (sherhut) [reply tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc:59]: Also, `itemA` is an alloc, so it cannot be a block argument, right?

File: tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+        while (insertionPoint != potReuseVector.end()) {
+          if (userange.getFirstUseIndex(itemB) <
+              userange.getFirstUseIndex(*insertionPoint))
+            break;
+          ++insertionPoint;
+        }
+        potReuseVector.insert(insertionPoint, itemB);
+      }
+
+      potentialReuseMap.insert(
```
Comment (sherhut) [inline tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc:80]: Do you need the full type here or would `{itemA, potReuseVector}` suffice?

File: tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+  /// Checks if there is a transitive interference between potReuseValue and the
+  /// value that may replace it, we call this value V. potReuses is the vector
+  /// of all values that can potentially be replaced by V. If potReuseValue
+  /// already replaces any other value that is not part of the potReuses vector
+  /// it cannot be replaced by V anymore.
+  bool transitiveInterference(
+      Value potReuseValue, SmallVector<Value, 4> &potReuses,
+      llvm::MapVector<Value, DenseSet<Value>> &actualReuseMap) {
+    return actualReuseMap.find(potReuseValue) != actualReuseMap.end() &&
+           llvm::any_of(actualReuseMap[potReuseValue], [&](Value vReuse) {
```
Comment (sherhut) [inline tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc:113]: Nit: This does a double lookup. If you already use find, you can use the result of the find to access the found value.

File: tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+             return !std::count(potReuses.begin(), potReuses.end(), vReuse);
```
Comment (sherhut) [inline tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc:114]: Nit: Use find here, as that stops after first occurrence.

File: tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+    Operation *defOpB = b.getDefiningOp();
+
+    // If the alloc method or the number of operands is not the same the types
+    // might not be compatible.
+    if (defOpA->getName() != defOpB->getName() ||
+        defOpA->getNumOperands() != defOpB->getNumOperands())
+      return false;
+
+    // If all operands are equal the types are compatible.
+    for (auto const &pair :
```
Comment (sherhut) [inline tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc:144]: This is not true. Consider `memref<?x5xi32>` vs. `memref<5x?xi32`. Also, you need to consider the basetype here, too.

As an extension, this could also work on the size of the element type. Not in this PR, please leave a TODO for this.

File: tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc [hunk 14]
```diff
@@ -0,0 +1,297 @@
...
+      if (std::get<0>(pair) != std::get<1>(pair))
+        return false;
+    }
+    return true;
+  }
+
+  /// A Fixpoint iteration over the potential reuses to compute the actual
+  /// reuses.
+  llvm::MapVector<Value, DenseSet<Value>> computeActualReuse(
+      llvm::MapVector<Value, SmallVector<Value, 4>> &potentialReuseMap) {
```
Comment (sherhut) [inline tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc:155]: Having read this type so often, maybe give it a name?

File: tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc [hunk 15]
```diff
@@ -0,0 +1,297 @@
...
+  /// For each value in the potentialReuseMap, check if another value tries to
+  /// reuse it or if it is already replaced by another value. If neither is the
+  /// case add the value and its reuses (if any) to the actualReuseMap.
+  void choosePotentialReuses(
+      DenseSet<Value> &replacedSet, DenseSet<Value> &currentReuserSet,
+      llvm::MapVector<Value, SmallVector<Value, 4>> &potentialReuseMap,
+      llvm::MapVector<Value, DenseSet<Value>> &actualReuseMap) {
+    for (auto &potReuser : potentialReuseMap) {
+      Value item = potReuser.first;
+      SmallVector<Value, 4> potReuses = potReuser.second;
```
Comment (sherhut) [inline tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc:199]: nit: this copies the vector.

**Review by sherhut — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc (hunk 13 above)
Comment (sherhut) [reply tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc:145]: I was a bit imprecise. Please fix the issue with partially static types and also consider the basetype.

An extension to support reuse for elementtypes of the same size can be a TODO.

**Review by sherhut — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc [hunk 16]
```diff
@@ -0,0 +1,311 @@
...
+    Operation *defOpA = a.getDefiningOp();
+    Operation *defOpB = b.getDefiningOp();
+
+    // If the alloc method or the number of operands is not the same the types
+    // might not be compatible.
+    if (defOpA->getName() != defOpB->getName() ||
+        defOpA->getNumOperands() != defOpB->getNumOperands())
+      return false;
+
+    // TODO: Fix for memref<?x5xi32> vs memref<5x?xi32>, also consider the
```
Comment (sherhut) [inline tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc:157]: This is now fixed.

File: tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc (hunk 13 above)
Comment (sherhut) [reply tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_reuse.cc:160]: Use `std::equal`?

**Review by sherhut — CHANGES_REQUESTED** [review]
Why is the code formatted differently in this PR? Is that an accidental change?

Please fix the comment and use std::equal. I will run presubmits on this change anyway, though.

File: tensorflow/compiler/mlir/hlo/BUILD [hunk 17]
```diff
@@ -1693,6 +1694,24 @@ cc_library(
...
+    srcs = ["lib/Transforms/buffer_reuse.cc"],
+    hdrs = [
+        "include/mlir-hlo/Analysis/userange_analysis.h",
+        "include/mlir-hlo/Transforms/PassDetail.h",
+        "include/mlir-hlo/Transforms/passes.h",
+    ],
+    deps = [
+        ":hlo",
+        ":transforms_pass_inc_gen",
+        "@llvm-project//mlir:MemRefDialect",
```
Comment (sherhut) [inline tensorflow/compiler/mlir/hlo/BUILD:1708]: These are ordered incorrect. See [the builder log](https://source.cloud.google.com/results/invocations/023880b0-268d-42dd-ae99-d6b088e90f2a/targets/%2F%2Ftensorflow%2Ftools%2Fci_build:gen_ci_sanity_out/log).

**Review by sherhut — APPROVED** [review]
